<?xml version="1.0" encoding="utf-8" ?>

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.3.9: http://docutils.sourceforge.net/" />
<meta http-equiv="expires" content="Wed, 19 Feb 2003 08:00:00 GMT">
<title></title>
<link rel="stylesheet" href="style/welcome.css" type="text/css" />

        <script type="text/javascript" src="scripts/jquery.js"></script>

        <script type="text/javascript">
            var q = jQuery.noConflict();
            q(document).ready(function() { 
                q( "a[minsizehint]" ).click( function() {
                    if ( parent.handle_minwidth_hint ) {
                        parent.handle_minwidth_hint( q(this).attr( "minsizehint" ) );
                    }
                });
            });
        </script>
		<script src="scripts/jquery_wz_jsgraphics.js" type="text/javascript"></script>

</head>
<body>

<div class="document">
    <div class="main">
	<div class="gsuiteinfobox">
	    <p style="font-size: 11px; align:left; margin:1px; padding-bottom: 5px;">
	    <a href="welcome.html?tabOn=tab3">&lt;&lt;Back to main</a></p>
	    <b>
		    Note on GSuite Use-cases:
	    </b>
	    <p>
		    The power of the GSuite HyperBrowser is best demonstrated by use-cases that show how tools are combined to execute a full analysis scenario.
		    We provide four use-cases that cover different analysis scenarios:
	    </p>
		    
	    <ul>
	    
		    <li>
			    The main use-case <a href="../u/hb-superuser/p/use-case---icgc-somatic-mutations" target="_blank">
			    understanding the genomic distribution of somatic mutations</a> - Performs a high-throughput and fully reproducible analysis of the distribution of 
			    non-coding somatic cancer mutation data across cancer types from ICGC against DNase1 hypersensitive sites across cells from ENCODE. 
		    
		    </li>
		    <li>
			    <a target="_blank" href="https://hyperbrowser.uio.no/nar/u/hb-superuser/p/use-case---tfs-in-enhancers">
			    Finding the frequency with which a given TF binds to enhancer elements</a> - Answers the question of how often does a given TF 
			    (Nanog in these examples) bind enhancer elements in K562 human cells. 
			    The use-case has been designed in such a way that the user must get the data from an external source (in this case, UCSC-ENCODE), 
			    upload it to History, and process it from there. 
		    </li>
		    <li>
		    
			    <a target="_blank" href="https://hyperbrowser.uio.no/nar/u/hb-superuser/p/use-case---overlap-of-snp-platforms">
			    Inspecting overlap or lack of overlap for SNP microrray platforms</a> - Evaluates a few selected SNP-microarray platforms 
			    based on publicly available annotation. The GSuite is manually created and imported into HyperBrowser. The referenced tracks 
			    are then downloaded and preprocessed. The Venn Diagram tool is used to calcualte and visualize the overlap between the tracks in the collection.
		    </li>
		    <li>
			    <a target="_blank" href="https://hyperbrowser.uio.no/nar/u/hb-superuser/p/use-case---characterizing-mirna-loci-by-reference-collections">
			    Comparing miRNA targets with DNA binding proteins, somatic mutations and repeating elements</a> - Compiles reference genomic track collections of 
			    somatic mutations, repeating elements, histone modifications and DNA binding proteins, from ICGC, ENCODE and from files in a local history.
			    Then uses these reference collections to characterize a set of regions of interest (a single genomic track of miRNA loci). 
		    </li>
	    </ul>
	</div>
    </div>
</div>

<div class="contact">

    <b>Note:</b>
    <p>
    This is a specialized version of The Genomic Hyperbrowser,
    focusing on new functionality for analyzing collections of genomic track.
    More information on the general HyperBrowser functionality and project can be
    found on the <a href="https://hyperbrowser.uio.no/hb/" target="_blank">main version</a> of the HyperBrowser.
    </p>
    <b>Contact information</b>
    <p>
    If you have any questions, requests or comments of any kind, please use our discussion forum, available from the <a minsizehint="-1" href="https://sites.google.com/site/hyperbrowserhelp/?tool_id=hb_help">Help</a> menu, or send an email to <a href="#" onclick="this.href='mailto:'+ this.innerHTML.split('').reverse().join('')" class="codedirection">on.oiu.tisu@stseuqer-resworbrepyh</a>&#x200E;. We are happy to respond. We especially encourage collaborative projects for further development. If you find any bugs, please send an email to
  <a href="#" onclick="this.href='mailto:'+ this.innerHTML.split('').reverse().join('')" class="codedirection">on.oiu.tisu@sgub-resworbrepyh</a>.</p>
    <p>If you register as a user, you will get the option to subscribe to the <a href="#" onclick="this.href='mailto:'+ this.innerHTML.split('').reverse().join('')" class="codedirection">on.oiu.tisu@ofni-resworbrepyh</a> mailing list. Here, we will send announcements and administrative messages for users of the system. A backlog of announcements is available from the Help menu. Citation info is also available from the Help menu.</p>

</div>


<p class="blocktext">
This project is being developed by the Norwegian bioinformatics community as an open-source project under the GPL  license v3,
supported by various national and local bodies.
</p>
<p class="blocktext">
  For further info on Norwegian bioinformatics, see <a href="http://www.bioinfo.no">www.bioinfo.no</a>.
</p>

<div class="section">
  <img src="hyperbrowser/images/logo/HB_support.png" class="center">
</div>

</body>
